crates/sui/src/upgrade_compatibility.rs

```diff
@@ -378,36 +381,159 @@ pub(crate) async fn check_compatibility(
-        .iter()
+    let new_modules_map: HashMap<Identifier, CompiledModule> = new_package
+        .get_modules()
         .map(|m| (m.self_id().name().to_owned(), m.clone()))
         .collect();
 
-    // for each existing find the new one run compatibility check
-    for existing_module in existing_modules {
-        let name = existing_module.self_id().name().to_owned();
-
-        // find the new module with the same name
-        match new_modules_map.get(&name) {
-            Some(new_module) => {
-                Compatibility::upgrade_check().check_with_mode::<CliCompatibilityMode>(
-                    &Module::new(&existing_module),
-                    &Module::new(new_module),
-                )?;
+    let errors: Vec<(Identifier, UpgradeCompatibilityModeError)> = existing_modules
+        .iter()
+        .flat_map(|existing_module| {
+            let name = existing_module.self_id().name().to_owned();
+
+            // find the new module with the same name
+            match new_modules_map.get(&name) {
+                Some(new_module) => {
+                    let compatible = Compatibility::upgrade_check()
+                        .check_with_mode::<CliCompatibilityMode>(
+                            &Module::new(existing_module),
+                            &Module::new(new_module),
+                        );
+                    if let Err(errors) = compatible {
+                        errors.into_iter().map(|e| (name.to_owned(), e)).collect()
+                    } else {
+                        vec![]
+                    }
+                }
+                None => vec![(
+                    name.clone(),
+                    UpgradeCompatibilityModeError::ModuleMissing { name },
+                )],
             }
+        })
+        .collect();
+
+    if errors.is_empty() {
+        return Ok(());
+    }
+
+    let mut files = SimpleFiles::new();
+    let config = codespan_reporting::term::Config::default();
+    let mut writer;
+    if enable_colors {
+        writer = codespan_reporting::term::termcolor::Buffer::ansi();
+    } else {
+        writer = codespan_reporting::term::termcolor::Buffer::no_color();
+    }
+    let mut file_id_map = HashMap::new();
+
+    for (name, err) in errors {
+        let compiled_unit_with_source = new_package
+            .package
+            .get_module_by_name_from_root(&name.to_string())
+            .context("Unable to get module")?;
+
+        let source_path = compiled_unit_with_source.source_path.to_string_lossy();
+        let file_id = match file_id_map.get(&source_path) {
+            Some(file_id) => *file_id,
             None => {
-                Err(anyhow!("Module {} is missing from the package", name))?;
+                let source = fs::read_to_string(&compiled_unit_with_source.source_path)
+                    .context("Unable to read source file")?;
+                let file_id = files.add(source_path.clone(), source);
```

Comment:
It's fine to use the path after lossy conversion as an input to `SimpleFiles` but we should not use it as the key to our map here, because we could get unintended name collisions that way -- we should be able to use the original path for the key.